### PR TITLE
gh-92154: Expose PyCode_GetCode in the C API

### LINKED ISSUE #92154: [C-API] Add a function to access PyCodeObject.co_code in C
**Feature or enhancement**

Previously, you could access the `co_code` field in `PyCodeObject` directly to access the bytecode. This is no longer the case since https://github.com/python/cpython/commit/2bde6827ea4f136297b2d882480b981ff26262b6. Currently in 3.11a7+, C code cannot easily obtain the equivalent of `co_code` in 3.10. `co_code_adaptive` gets quickened bytecode which is not equivalent to the 3.10 behavior.

I would imagine many tools relying on the old behavior will break.

**Pitch**

I propose we expose some form of the currently internal `_PyCode_GetCode`.

**Previous discussion**

Related to #91397.
See also coverage.py issue https://github.com/nedbat/coveragepy/issues/1367

CC @vstinner @nedbat @brandtbucher @markshannon

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
Fixes #92154.

### CODE REVIEW

File: Objects/codeobject.c [hunk 1]
```diff
@@ -1370,6 +1370,15 @@ _PyCode_GetCode(PyCodeObject *co)
     return code;
 }
 
+PyObject *
+PyCode_GetCode(PyObject *co)
+{
+    if (!PyCode_Check(co)) {
+        PyErr_BadInternalCall();
```
Comment (Fidget-Spinner) [inline Objects/codeobject.c]: I forgot if this or a `TypeError` is preferred for the C API (or no error setting at all?)

**Comment by Fidget-Spinner** [comment]
Also I frankly have no clue if we need to update `stable_abi.toml`. I'm not sure when it counts as stable ABI and when it doesn't.

File: Objects/codeobject.c (hunk 1 above)
Comment (markshannon) [reply Objects/codeobject.c]: Use `PyCode_GetCode(PyCodeObject *co)` and then you won't need the check.

Comment (Fidget-Spinner) [reply Objects/codeobject.c]: I would like to, but `co_code` was previously `PyObject*`, and I want this to be a drop-in replacement as far as possible.

File: Doc/c-api/code.rst [hunk 2]
```diff
@@ -76,3 +76,16 @@ bound into a function.
    information is not available for any particular element.
 
    Returns ``1`` if the function succeeds and 0 otherwise.
+
+.. c:function:: PyObject* PyCode_GetCode(PyObject *co)
+
+   Equivalent to the Python code ``getattr(co, 'co_code')``.
+   Returns a strong reference to a :c:type:`PyBytesObject` representing the
+   bytecode in a code object. On error, ``NULL`` is returned.
```
Comment (vstinner) [inline Doc/c-api/code.rst]:
```suggestion
   bytecode in a code object. On error, ``NULL`` is returned and raise an exception.
```

File: Doc/whatsnew/3.11.rst [hunk 3]
```diff
@@ -1408,6 +1408,10 @@ C API Changes
   To get a custom code object: create a code object using the compiler,
   then get a modified version with the ``replace`` method.
 
+* :c:type:`PyCodeObject` no longer has a ``co_code`` field.  Instead,
+  use ``PyObject_GetAttrString(code_object, 'co_code')`` or
+  :c:func:`PyCode_GetCode` to get the underlying bytes object.
```
Comment (vstinner) [inline Doc/whatsnew/3.11.rst:1413]: Great! Thanks.

File: Doc/whatsnew/3.11.rst [hunk 4]
```diff
@@ -1408,6 +1408,10 @@ C API Changes
   To get a custom code object: create a code object using the compiler,
   then get a modified version with the ``replace`` method.
 
+* :c:type:`PyCodeObject` no longer has a ``co_code`` field.  Instead,
+  use ``PyObject_GetAttrString(code_object, 'co_code')`` or
```
Comment (vstinner) [inline Doc/whatsnew/3.11.rst]: In C, single quote is for single char: ``'x'``

```suggestion
  use ``PyObject_GetAttrString(code_object, "co_code")`` or
```

File: Doc/whatsnew/3.11.rst [hunk 5]
```diff
@@ -1408,6 +1408,10 @@ C API Changes
...
+  :c:func:`PyCode_GetCode` to get the underlying bytes object.
+
```
Comment (vstinner) [inline Doc/whatsnew/3.11.rst:1415]: I may help to add a link to the issue "Contributed by ... issue ...". I like to go to the issue to get the rationale of a change, for example.

File: Objects/codeobject.c (hunk 1 above)
Comment (vstinner) [reply Objects/codeobject.c]: PyErr_BadInternalCall() is fine, but I would even suggest replacing a runtime check with an assertion. The caller is responsible to pass the right type.

**Comment by vstinner** [comment]
> Also I frankly have no clue if we need to update stable_abi.toml. I'm not sure when it counts as stable ABI and when it doesn't.

Include/cpython/ is the API excluded from the limited C API and so excluded from the stable ABI: https://devguide.python.org/c-api/

Please don't add this function to the stable ABI yet. Let's wait for one Python release, and then see if it's stable or not.

File: Doc/c-api/code.rst [hunk 6]
```diff
@@ -76,3 +76,16 @@ bound into a function.
    information is not available for any particular element.
 
    Returns ``1`` if the function succeeds and 0 otherwise.
+
+.. c:function:: PyObject* PyCode_GetCode(PyObject *co)
```
Comment (vstinner) [inline Doc/c-api/code.rst]: The two API giving code objects return type ``PyCodeObject*``:

* PyCode_NewEmpty()
* PyFrame_GetCode()

So it sounds more consistent to use ``PyCodeObject*`` for the parameter.

**Comment by vstinner** [comment]
I prepared PR https://github.com/python/pythoncapi-compat/pull/34 to add the function to pythoncapi-compat.

**Review by vstinner — APPROVED** [review]
LGTM.

**Pull request merged**